### PR TITLE
change the inference context api to use TypeProto

### PR DESCRIPTION
**Pull request by linkerzhang**

### CODE REVIEW

File: onnx/defs/shape_inference.h [hunk 1]
```diff
@@ -44,17 +45,43 @@ inline bool hasExactlyNInputTypes(InferenceContext& ctx, int n, const std::strin
   return true;
 }
 
-inline void propagateElemTypeFromInputToOutput(InferenceContext& ctx, size_t inputIndex, size_t outputIndex) {
-  ctx.getOutputType(outputIndex)->set_elem_type(ctx.getInputType(inputIndex)->elem_type());
+inline void propagateElemTypeFromInputToOutput(
```
Comment (gramalingam) [inline onnx/defs/shape_inference.h]: Change name to "propagateTypeFromInputToOutput"?

**Comment by bddppq** [comment]
Could you only clang-format the changed parts? :-) It's pretty hard to review.

File: onnx/shape_inference/implementation.h [hunk 2]
```diff
@@ -140,18 +148,21 @@ void InferShapes(ModelProto& m) {
...
+                  "inferred dimension differs from existing dimension");
             }
             *existingDim = inferredDim;
           }
         }
       }
 
       // Make it available to futher inference.
-      valueTypesByName[output] = existingType;
+      valueTypesByName[output] = iter->second;
```
Comment (bddppq) [inline onnx/shape_inference/implementation.h]: what if `iter == valueTypesByName.end()`?

File: onnx/defs/shape_inference.h (hunk 1 above)
Comment (linkerzhang) [reply onnx/defs/shape_inference.h]: I found that what it was supposed to do is really coping element type instead of whole type proto.

**Comment by bddppq** [comment]
@anderspapitto Could you take a look?

**Pull request merged**